src/mongoose_cluster_id.erl

```diff
@@ -0,0 +1,166 @@
+-module(mongoose_cluster_id).
+-compile([export_all]).
+
+-include("mongoose.hrl").
+
+-export([
+         start/0,
+         get_cluster_id/0,
+         get_backend_cluster_id/0
+        ]).
+
+-export([which_backend_available/0]).
+
+-record(mongoose_cluster_id, {key :: atom(), value :: cluster_id()}).
+-type cluster_id() :: binary().
+-type mongoose_backend() :: rdbms
+                          % | cassandra
+                          % | elasticsearch
+                          % | riak
+                          | mnesia.
+
+-spec start() -> ok | {error, any()}.
+start() ->
+    init_mnesia_cache(),
+    case get_cluster_id() of
+        {error, _} = E ->
+            ?WARNING_MSG("get_cluster_id failed ~p", [E]),
+            case get_backend_cluster_id() of
+                {error, _} = E ->
+                    ?WARNING_MSG("get_backend_cluster_id failed ~p", [E]),
+                    NewID = make_cluster_id(),
+                    case set_new_cluster_id(NewID) of
+                        {error, _} = Err ->
+                            ?ERROR_MSG("Everything in cluster_id failed! ~p", [Err]),
+                            Err;
+                        ok -> cache_cluster_id(NewID)
+                    end;
+                {ok, ID} when is_binary(ID) ->
+                    cache_cluster_id(ID)
+            end;
+        {ok, ID} when is_binary(ID) -> ok
+    end.
+
+%% Get cached version
+-spec get_cluster_id() -> {ok, cluster_id()} | {error, any()}.
+get_cluster_id() ->
+    T = fun() -> mnesia:read(mongoose_cluster_id, cluster_id) end,
+    case mnesia:transaction(T) of
+        {atomic, [#mongoose_cluster_id{value = ClusterID}]} ->
+            {ok, ClusterID};
+        {atomic, []} ->
+            {error, cluster_id_not_created};
+        {aborted, Reason} ->
+            {error, Reason}
+    end.
+
+-spec get_backend_cluster_id() -> {ok, cluster_id()} | {error, any()}.
+get_backend_cluster_id() ->
+    get_cluster_id_from_backend(which_backend_available()).
+
+%% ====================================================================
+%% Internal functions
+%% ====================================================================
+init_mnesia_cache() ->
+    mnesia:create_table(mongoose_cluster_id,
+                        [{type, set},
+                         {record_name, mongoose_cluster_id},
+                         {attributes, record_info(fields, mongoose_cluster_id)},
+                         {disc_only_copies, [node()]}
```

Comment:
Did you test (manually) if this `disc_only_copies` causes a dead lock in restarting clustered nodes? Possible cases:
1. Start 2 nodes and cluster them together. Stop both of them, start the second first (see if it starts properly - is able to accept connections) and the start the first one and check if everything is ok.
2. Similar test with 3 nodes and different starting / re-starting order.

Reply:
Just tested now (for some miraculous reason I can test this manually considerably faster in my private laptop compare to the company one :joy: ). Bad news, it doesn't seem to work. When stopping both and then starting the second, this second does not accept connections. What the heck is wrong with mnesia :disappointed:

Reply:
Although, actually, the second node does not connect anything when doing these steps, even when I put everything as `ram_copies`. This reproduces to me even in master :no_mouth: